File: frontend/src/types/execution.ts
/**
 * Types for code execution API.
 */

export interface ExecutionRequest {
  code: string;
  timeout_seconds?: number;
}

export interface AsyncExecutionRequest {
  code: string;
  connection_id: string;
  timeout_seconds?: number;
}

export interface JobSubmittedResponse {
  job_id: string;
  status: string;
}

export interface ExecutionResponse {
  success: boolean;
  stdout: string;
  stderr: string;
  error: string | null;
  error_type: string | null;
  execution_time_ms: number;
  timed_out: boolean;
  security_violations: SecurityViolation[];
}

export interface SecurityViolation {
  type: string;
  message: string;
  line: number | null;
  column: number | null;
}

/**
 * Types for code analysis API.
 */

export interface AnalyzeRequest {
  code: string;
  snippet_id?: string;
}

export interface AsyncAnalyzeRequest {
  code: string;
  connection_id: string;
  snippet_id?: string;
}

export interface AnalyzeResponse {
  success: boolean;
  time_complexity: string;
  space_complexity: string;
  narrative: string;
  error: string | null;
  available: boolean;
  model: string | null;
}

export interface AnalysisStatus {
  available: boolean;
  provider: string | null;
}

/**
 * WebSocket message types for analysis streaming.
 */

export interface AnalysisStreamStartMessage {
  type: 'analysis_stream_start';
  job_id: string;
}

export interface AnalysisStreamChunkMessage {
  type: 'analysis_stream_chunk';
  job_id: string;
  chunk: string;
}

export interface AnalysisStreamCompleteMessage {
  type: 'analysis_stream_complete';
  job_id: string;
  result: AnalyzeResponse;
}

export interface AnalysisStreamErrorMessage {
  type: 'analysis_stream_error';
  job_id: string;
  error: string;
}

File: backend/api/schemas/execution.py
"""Pydantic schemas for code execution."""

from pydantic import BaseModel, Field


class ExecutionRequest(BaseModel):
    """Request to execute Python code."""

    code: str = Field(
        ...,
        min_length=1,
        max_length=10240,
        description="Python code to execute (max 10KB)",
    )
    timeout_seconds: float = Field(
        default=30.0,
        gt=0,
        le=30.0,
        description="Maximum execution time in seconds (max 30s)",
    )


class AsyncExecutionRequest(BaseModel):
    """Request to execute Python code asynchronously with WebSocket callback."""

    code: str = Field(
        ...,
        min_length=1,
        max_length=10240,
        description="Python code to execute (max 10KB)",
    )
    connection_id: str = Field(
        ...,
        min_length=1,
        max_length=256,
        description="WebSocket connection ID for result delivery",
    )
    timeout_seconds: float = Field(
        default=30.0,
        gt=0,
        le=30.0,
        description="Maximum execution time in seconds (max 30s)",
    )


class JobSubmittedResponse(BaseModel):
    """Response when an async job is submitted."""

    job_id: str = Field(..., description="Unique job identifier")
    status: str = Field(default="queued", description="Job status")


class SecurityViolationResponse(BaseModel):
    """A security violation found in the code."""

    line: int
    column: int
    message: str


class ExecutionResponse(BaseModel):
    """Response from code execution."""

    success: bool = Field(..., description="Whether execution completed successfully")
    stdout: str = Field(default="", description="Standard output from execution")
    stderr: str = Field(default="", description="Standard error from execution")
    error: str | None = Field(default=None, description="Error message if execution failed")
    error_type: str | None = Field(default=None, description="Type of error (e.g., ValueError)")
    execution_time_ms: float = Field(default=0.0, description="Execution time in milliseconds")
    timed_out: bool = Field(default=False, description="Whether execution timed out")
    security_violations: list[SecurityViolationResponse] = Field(
        default_factory=list,
        description="Security violations found in the code",
    )
